core/src/main/scala/cats/data/NonEmptyVector.scala

```diff
@@ -0,0 +1,230 @@
+
+      override def reduceLeft[A](fa: NonEmptyVector[A])(f: (A, A) => A): A =
+        fa.reduceLeft(f)
+
+      override def reduce[A](fa: NonEmptyVector[A])(implicit A: Semigroup[A]): A =
+        fa.reduce
+
+      override def map[A, B](fa: NonEmptyVector[A])(f: A => B): NonEmptyVector[B] =
+        fa map f
+
+      def pure[A](x: A): NonEmptyVector[A] =
+        NonEmptyVector(x, Vector.empty)
+
+      def flatMap[A, B](fa: NonEmptyVector[A])(f: A => NonEmptyVector[B]): NonEmptyVector[B] =
+        fa flatMap f
+
+      def coflatMap[A, B](fa: NonEmptyVector[A])(f: NonEmptyVector[A] => B): NonEmptyVector[B] = {
+        @tailrec def consume(as: Vector[A], buf: VectorBuilder[B]): Vector[B] =
+          as match {
+            case a +: as => consume(as, buf += f(NonEmptyVector(a, as)))
+            case _ => buf.result()
+          }
+        NonEmptyVector(f(fa), consume(fa.tail, new VectorBuilder[B]))
+      }
+
+      def extract[A](fa: NonEmptyVector[A]): A = fa.head
+
+      def traverse[G[_], A, B](fa: NonEmptyVector[A])(f: (A) => G[B])(implicit G: Applicative[G]): G[NonEmptyVector[B]] =
+      G.map2Eval(f(fa.head), Always(Traverse[Vector].traverse(fa.tail)(f)))(NonEmptyVector(_, _)).value
+
+
+      override def foldLeft[A, B](fa: NonEmptyVector[A], b: B)(f: (B, A) => B): B =
+        fa.foldLeft(b)(f)
+
+      override def foldRight[A, B](fa: NonEmptyVector[A], lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B] =
+        fa.foldRight(lb)(f)
+
+      def tailRecM[A, B](a: A)(f: A => NonEmptyVector[A Xor B]): NonEmptyVector[B] = {
+        val buf = new VectorBuilder[B]
+        @tailrec def go(v: NonEmptyVector[A Xor B]): Unit = v.head match {
+            case Xor.Right(b) =>
+            buf += b
+            NonEmptyVector.fromVector(v.tail) match {
+              case Some(t) => go(t)
+              case None => ()
+            }
+          case Xor.Left(a) => go(f(a).concat(v.tail))
+          }
+        go(f(a))
+        NonEmptyVector.fromVectorUnsafe(buf.result())
+      }
+    }
+
+  implicit def catsDataEqForNonEmptyVector[A](implicit A: Eq[A]): Eq[NonEmptyVector[A]] =
+    new Eq[NonEmptyVector[A]]{
+      def eqv(x: NonEmptyVector[A], y: NonEmptyVector[A]): Boolean = x === y
+    }
+
+  implicit def catsDataShowForNonEmptyVector[A](implicit A: Show[A]): Show[NonEmptyVector[A]] =
+    Show.show[NonEmptyVector[A]](_.show)
+
+  implicit def catsDataSemigroupForNonEmptyVector[A]: Semigroup[NonEmptyVector[A]] =
+    catsDataInstancesForNonEmptyVector.algebra
+
+}
+
+object NonEmptyVector extends NonEmptyVectorInstances {
+
+  def apply[A](head: A, tail: Vector[A]): NonEmptyVector[A] =
+    NonEmptyVector(head +: tail)
+
+  def apply[A](head: A, tail: A*): NonEmptyVector[A] = {
+    val buf = Vector.newBuilder[A]
+    buf += head
+    tail.foreach(buf += _)
+    NonEmptyVector(buf.result)
+  }
+
+  def fromVector[A](vector: Vector[A]): Option[NonEmptyVector[A]] =
+    vector.headOption.map(apply(_, vector.tail))
```

Comment:
+1. I think we should do the horrible, but fast stuff internally when we can prove it is correct (and then test it to be sure).